### DISCUSSION TITLE
Formatting python with using pyright LSP and black

### DISCUSSION THREAD

**Discussion by speri203** [LazyVim/LazyVim#257] [General]
I just switched over to LazyVim today and have it pretty much configured how I like it. The only issue I am facing is utilizing the black formatter to format my code on save. When pyright + black is installed and I do <leader>cf, a notification pops up saying "_[LSP] request failed no matching language server_" but when I uninstall pyright and use the python-lsp-server then the formatting works but with autopep8. Is there a way I can utilize pyright and black to do this instead?

**Comment by dpetka2001** [comment]
I don't know how you installed `black`, but I installed it with `mason` and configured it in `null-ls`, so that it uses `black` for formatting. It's as easy as following:
```lua
  -- Add `pyright` to mason
  {
    "williamboman/mason.nvim",
    opts = function(_, opts)
      vim.list_extend(opts.ensure_installed, { "pyright", "black" })
    end,
  },

  -- Setup null-ls with `black`
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, { nls.builtins.formatting.black })
    end,
  },
```
Then if you do `NullLsInfo` in a python file you should see that `black` formatter should be attached.

**Reply by speri203** [reply]
Yes, this works. I was using mason to install all my lsp related stuff. Thank you!

**Reply by fredrikaverpil** [reply]
I had the same issue as @speri203 but when using the above config posted by @dpetka2001 I consistently get a neovim crash when executing `:wqa`. This is when I have a bunch of unsaved, changed buffers, following a symbol rename.

I can have ruff and mypy added like this, but not black for some reason.

```lua
  -- change null-ls config
  {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local null_ls = require("null-ls")

      -- temporarily disabled the below as it crashes Neovim on e.g. `<leader>cr` followed by `:wqa`.
      -- -- add black as formatter
      -- opts.sources = vim.list_extend(opts.sources, { null_ls.builtins.formatting.black })

      -- add mypy as linter
      opts.sources = vim.list_extend(opts.sources, { null_ls.builtins.diagnostics.mypy })

      -- add ruff as linter
      opts.sources = vim.list_extend(opts.sources, { null_ls.builtins.diagnostics.ruff })

      -- remove flake8 from opts.sources
      opts.sources = vim.tbl_filter(function(source)
        return source.name ~= "flake8"
      end, opts.sources)
    end,
  },
```

@speri203 do you see this issue too with black and when saving a bunch of unsaved and changed buffers, e.g. following a symbol rename (`<leader>cr`)?

**Reply by dpetka2001** [reply]
What about if you first `wa` and then `q` to quit? Does it still happen then?

**Reply by speri203** [reply]
Yeah I was having the issue with crashes as well but even with only one buffer open. I was able to remedy it with this but my null-ls looks like this

 ```
 {
    "jose-elias-alvarez/null-ls.nvim",
    opts = function(_, opts)
      local nls = require("null-ls")
      opts.sources = vim.list_extend(opts.sources, { nls.builtins.formatting.black })
    end,
  },
```

This consistently formats on save. Hopefully this helps you as well @fredrikaverpil 
EDIT: I am using flake8 as my linter and pyright as my LSP.
I will test this out today to see if I experience the same issue you are having.

**Comment by fredrikaverpil** [comment]
> What about if you first `wa` and then `q` to quit? Does it still happen then?

It's the `:wa` that crashes neovim for me. I suspect something goes wrong when it's saving multiple files and also applying formatting.

If I hit `<leader>,` and save each file individually, there's no issue.

**Reply by dpetka2001** [reply]
Unfortunately I'm not familiar with this behavior. I tried saving multiple python files open in buffers with formatting and my Neovim didn't crash. I don't however have symbols that expand in multiple files. So if I rename a symbol it only gets renamed within the same file. I'm a beginner student and don't write too much complicated code. Maybe the problem lies with renaming symbols that expand in more than one file(if that is your situation)? Honestly I don't know what to tell you as I haven't encountered it yet. Have you tried a different formatter(maybe autopep8 or yapf) to see if the problem persists?

**Comment by fredrikaverpil** [comment]
The crash is gone for me after installing the nightly 0.9.x-build: https://github.com/neovim/neovim/releases/tag/nightly

**Comment by fredrikaverpil** [comment]
For what it's worth, I got pyright + black (and also ruff and mypy) set up and I opted for a bit of a custom setup.

I want to install LSPs via Mason and have these centrally installed. But for ruff and black mypy I want them in my project's virtual environment. So I've opted for a bit of a mixed setup where I will tell null-ls to expect to find some stuff on `$PATH` and some stuff in the Mason installation path.

```lua
return {

  -- change mason config
  -- note: don't forget to update treesitter for languages
  {
    "williamboman/mason.nvim",
    -- opts will be merged with the parent spec
    opts = {
      ensure_installed = {
        -- python
        "ruff-lsp",
        "pyright",

        -- lua
        "lua-language-server",
        "stylua",

        -- shell
        "shellcheck",
        "shfmt",

        -- docker
        "dockerfile-language-server",

        -- javascript/typescript
        "prettierd",
        "typescript-language-server",
        "eslint-lsp",

        -- rust
        "rustfmt",
        "rust-analyzer",

        -- go
        -- handled by lazy.lua
      },
    },
  },

  -- change null-ls config
  {
    "jose-elias-alvarez/null-ls.nvim",
    dependencies = { "mason.nvim" },
    event = { "BufReadPre", "BufNewFile" },
    opts = function()
      local mason_registry = require("mason-registry")
      local null_ls = require("null-ls")
      local formatting = null_ls.builtins.formatting
      local diagnostics = null_ls.builtins.diagnostics
      local code_actions = null_ls.builtins.code_actions

      null_ls.setup({
        -- debug = true, -- Turn on debug for :NullLsLog
        debug = false,
        -- diagnostics_format = "#{m} #{s}[#{c}]",
        sources = {
          -- list of supported sources:
          -- https://github.com/jose-elias-alvarez/null-ls.nvim/blob/main/doc/BUILTINS.md

          -- get from $PATH
          diagnostics.ruff,
          diagnostics.mypy,
          formatting.black,

          -- get from mason
          formatting.stylua.with({
            command = mason_registry.get_package("stylua").path,
            extra_args = { "--indent-type", "Spaces", "--indent-width", "2" },
          }),
          formatting.shfmt.with({
            command = mason_registry.get_package("shfmt").path,
          }),
          formatting.prettierd.with({
            command = mason_registry.get_package("prettierd").path,
          }),
          formatting.rustfmt.with({
            command = mason_registry.get_package("rustfmt").path,
          }),
          formatting.yamlfix.with({
            command = mason_registry.get_package("yamlfix").path, -- requires python
          }),

          diagnostics.yamllint.with({
            command = mason_registry.get_package("yamllint").path,
          }),

          code_actions.shellcheck.with({
            command = mason_registry.get_package("shellcheck").path,
          }),
        },
      })
    end,
  },
}
```

Please note that ruff-lsp is needed for organization of imports etc via `<leader>ca` (code actions).


Here's the actual file since this tends to change over time 😉 
https://github.com/fredrikaverpil/dotfiles/blob/main/nvim-lazyvim/lua/plugins/lsp.lua

**Comment by slavanorm** [comment]
I am currently struggling with running < leader > cf / code format, getting error "No formatter available".

I have problems with mason setting up my python lsp, so i do not configure it.
I am assuming that i should configure the imports inside the nvim-lsp setup function.

update:
configuration ended very easily:
i disabled lang.python in LazyExtras,
and manually set opts for conform.nvim.

< leader > cf is still working for me.

Thanks so much for allowing to customize lazyvim into any direction =)
